vos/vos/storage_inventory.py

```diff
@@ -0,0 +1,315 @@
+            if send_md5:
+                return metadata.get('content_md5')
+            elif disposition:
+                return metadata.get('content_disposition')
+            else:
+                return metadata.get('content_length')
+        else:
+            raise ValueError('Unable to process copying {} to {}'.format(
+                                source, destination))
+
+    def delete(self, uri):
+        """Delete the Artifact
+        :param uri: The Artifact URI to delete.
+
+        :raises When a network problem occurs, it raises one of the
+        HttpException exceptions declared in the
+        cadcutils.exceptions module
+        """
+        logger.debug("delete {0}".format(uri))
+        url = '{}/{}'.format(
+            self._get_ws_client()._get_url((STANDARD_ID, None)), uri)
+        response = self.conn.session.delete(url)
+        response.raise_for_status()
+
+    def transfer(self, uri, direction, view=None, cutout=None):
+        transfer_url = '{}/{}'.format(
+            self._get_ws_client()._get_url((STANDARD_ID, None)), uri)
+        trans = Transfer()
+        return trans.transfer(transfer_url, uri, direction, self.conn,
+                              self._get_protocol(), view, cutout)
+
+    def get_transfer_error(self, uri, url):
+        trans = Transfer()
+        return trans.get_transfer_error(self.conn, uri, url)
+
+    def _get(self, source, destination):
+        """
+        :param source: The source Artifact to download (URI).
+        :type uri:  unicode
+        :param destination: The destination to write the data to (path).
+        :type destination: unicode
+        :return: destination size in bytes
+        :rtype long
+        """
+        source_md5 = self.get_metadata(source).get('content_md5')
+        download_url = '{}/{}'.format(
+            self._get_ws_client()._get_url((STANDARD_ID, None)), source)
+
+        stream = Stream(self.conn)
+        logger.debug('Downloading {}'.format(source))
+        return stream.download(download_url, source, source_md5, destination,
+                               True, True)
+
+    def _put(self, source, destination):
+        """
+        Perform the upload and verify the MD5 post write.
+
+        :param source: The source Artifact to download (URI).
+        :type uri:  unicode
+        :param destination: The destination to write the data to (path).
+        :type destination: unicode
+        :return: metadata dict for the artifact.
+        :rtype {}
+        """
+        source_md5 = md5_cache.MD5Cache.compute_md5(source)
+
+        upload_url = '{}/{}'.format(
+            self._get_ws_client()._get_url((STANDARD_ID, None)), destination)
+        stream = Stream(self.conn)
+        logger.debug('Uploading {}'.format(source))
+        return stream.upload(upload_url, destination, source, source_md5,
+                             self.get_metadata)
+
+    def _get_protocol(self):
+        if self.secure_get:
```

Comment:
We're just dealing with an issue right now where HTTPS is faster than HTTP. The conclusion was to have clients send the protocols they support (in this case both https and http) and let the server decide which one to pick.